src/compute/src/extensions/arrange.rs

```diff
@@ -0,0 +1,496 @@
+    T: Lattice + Timestamp,
+    R: Semigroup,
+{
+    fn log_arrangement_size(&self) {
+        log_arrangement_size_inner(&self.arranged, |trace| {
+            let (mut size, mut capacity, mut allocations) = (0, 0, 0);
+            let mut callback = |siz, cap| {
+                allocations += 1;
+                size += siz;
+                capacity += cap
+            };
+            trace.map_batches(|batch| {
+                batch.layer.keys.heap_size(&mut callback);
+                batch.layer.vals.keys.heap_size(&mut callback);
+                vec_size(&batch.layer.offs, &mut callback);
+                vec_size(&batch.layer.vals.offs, &mut callback);
+                vec_size(&batch.layer.vals.vals.vals, &mut callback);
+            });
+            (size, capacity, allocations)
+        });
+    }
+}
+
+impl<G, K, T, R> ArrangementSize for MzArranged<G, TraceAgent<RowKeySpine<K, T, R>>>
+where
+    G: Scope<Timestamp = T>,
+    G::Timestamp: Lattice + Ord,
+    K: Data + Columnation,
+    T: Lattice + Timestamp,
+    R: Semigroup,
+{
+    fn log_arrangement_size(&self) {
+        log_arrangement_size_inner(&self.arranged, |trace| {
+            let (mut size, mut capacity, mut allocations) = (0, 0, 0);
+            let mut callback = |siz, cap| {
+                allocations += 1;
+                size += siz;
+                capacity += cap
+            };
+            trace.map_batches(|batch| {
+                batch.layer.keys.heap_size(&mut callback);
+                vec_size(&batch.layer.offs, &mut callback);
+                vec_size(&batch.layer.vals.vals, &mut callback);
+            });
+            (size, capacity, allocations)
+        });
+    }
+}
+
+impl<G, V, T> ArrangementSize for ArrangementImport<G, V, T>
+where
+    G: Scope,
+    G::Timestamp: Lattice,
+    G::Timestamp: Refines<T>,
+    T: Timestamp + Lattice,
+    V: Data + Columnation,
+{
+    fn log_arrangement_size(&self) {
+        // Imported traces are logged during export, so we don't have to log anything here.
+    }
+}
+
+impl<G, T> ArrangementSize for ErrArrangementImport<G, T>
+where
+    G: Scope,
+    G::Timestamp: Lattice,
+    G::Timestamp: Refines<T>,
+    T: Timestamp + Lattice,
+{
+    fn log_arrangement_size(&self) {
+        // Imported traces are logged during export, so we don't have to log anything here.
+    }
+}
+
+/// A wrapper behaving similar to Differential's [`Arranged`] type.
+///
+/// This type enables logging arrangement sizes, but only if the trace is consumed downstream.
+#[derive(Clone)]
+pub struct MzArranged<G, Tr>
+where
+    G: Scope,
+    G::Timestamp: Lattice,
+    Tr: TraceReader<Time = G::Timestamp> + Clone,
+    Tr::Time: Lattice + Ord + Clone + 'static,
+{
+    /// The arrangement we're wrapping
+    arranged: Arranged<G, Tr>,
+    /// Shared boolean to indicate that someone is consuming the trace, and we've installed
+    /// the arrangement size logging operator.
+    trace_consumed: Rc<RefCell<bool>>,
```

Comment:
You could use `Rc<Cell<bool>>` instead, to get rid of an if-panic check when accessing the value.

Reply:
Yup, thanks!